### PR TITLE
 move leak-check to during coherence, candidate eval

### PR DESCRIPTION
**Pull request by nikomatsakis**
Implementation of MCP https://github.com/rust-lang/compiler-team/issues/295.

I'd like to do a crater run on this.

Note to @rust-lang/lang: This PR is a breaking change (bugfix). It causes tests like the following to go from a future-compatibility warning #56105 to a hard error:

```rust
trait Trait {}
impl Trait for for<'a, 'b> fn(&'a u32, &'b u32) {}
impl Trait for for<'c> fn(&'c u32, &'c u32) {} // now rejected, used to warn
```

I am not aware of any instances of this code in the wild, but that is why we are doing a crater run. The reason for this change is that those two types are, in fact, the same type, and hence the two impls are overlapping. 

There will still be impls that trigger #56105 after this lands, however -- I hope that we will eventually just accept those impls without warning, for the most part. One example of such an impl is this pattern, which is used by wasm-bindgen and other crates as well:

```rust
trait Trait {}
impl<T> Trait for fn(&T) { }
impl<T> Trait for fn(T) { } // still accepted, but warns
```

### CODE REVIEW

**Comment by nikomatsakis** [comment]
Whoops, I just noticed that I forgot to remove those first 2 commits from the history. I'll leave it as is for now so as not to disturb the try run, but we should clean that up before landing.

**Comment by Aaron1011** [comment]
@nikomatsakis: What would it mean to accept the impl

```rust
trait Trait {}
impl<T> Trait for fn(&T) { }
impl<T> Trait for fn(T) { } // still accepted, but warns
```

without warnings? Would the `impl<T> Trait for fn(&T) { }` essentialy be ignored, or would it implicitly 'specialize' `impl<T> Trait for fn(T) { }` without actually using `#![feature(specialization)]`)?

**Comment by nikomatsakis** [comment]
@Aaron1011 

Neither, they would be considered distinct types. In particular, the first type `fn(T)` cannot ever be equal to the second type `for<'a> fn(&'a U)`  because there is no value of `T` that can name `'a`. 

The reason for the warning, however, is that we cannot yet distinguish that type from those where implied bounds lets us figure out that `'a` is equal to some other region that `T` *could* name (I give some examples in the MCP, I think).

**Comment by nikomatsakis** [comment]
I can't tell if the try build succeeded or failed :) or maybe it's still pending?

**Comment by Aaron1011** [comment]
> Neither, they would be considered distinct types. In particular, the first type fn(T) cannot ever be equal to the second type for<'a> fn(&'a U) because there is no value of T that can name 'a.

I see. Does that mean that it's impossible to write a blanket impl (or collection of blanket impls) that apply to all potential `fn(T)`s? For example, neither `impl<T> Trait for fn(&T) { }` or `impl<T> Trait for fn(T) { }` would apply to `fn((&u8, &u8))`

**Comment by nikomatsakis** [comment]
@Aaron1011

> Does that mean that it's impossible to write a blanket impl (or collection of blanket impls) that apply to all potential `fn(T)`s? 

Yes, that is presently impossible. You might be interested though in [this proto-proposal](https://hackmd.io/lse14c98S4mgkne1IGiIBQ?view), which addesses some related use cases. (I've been wanting to try and move that forward, I was sort of waiting until we establish the new lang team process, though it's probably not necessary, I'd love to see someone experiment with it...)

**Comment by Mark-Simulacrum** [comment]
@craterbot run mode=check-only start=master#7f940ef5d91b53e889f111f27e00849f2f5ae4a2 end=try#e2aab98dd8975a0990f9fbb79b3ea4b40559a184

**Comment by lqf96** [comment]
@Aaron1011 In the long term I think this can be solved by introducing variadic lifetimes. For example, you would write something like `impl<'...a, T> Trait for fn(T<'...a>)` so that your implementation can apply to any type that takes zero or more lifetime parameters. The `'...a` lifetime parameter pack would capture all the lifetime parameters and their constraints of the `T` type.

**Comment by matthewjasper** [comment]
Regressions/fixes all look spurious.

**Comment by nikomatsakis** [comment]
Rebased.

**Comment by jplatte** [comment]
It seems very counter-intuitive to me to accept

```rust
trait Trait {}
impl<T> Trait for fn(&T) { }
impl<T> Trait for fn(T) { }
```

if the first impl doesn't apply to all references. Currently, it doesn't apply for `'static` references ([playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=6d39b548508eea352471bf640e010fa5)) and I don't think it could, otherwise there'd be an overlap.

**Comment by nikomatsakis** [comment]
@jplatte the trait is not being implemented for `&T`, it is being implemented for `fn(&T)`. The important point is whether these two types are the same:

* `for<'a> fn(&'a T)` -- a function that accepts references of any lifetime
* `fn(&'static T)` -- a function that accepts only `'static` references

Under the rules proposed here, and current behavior, we can indeed distinguish between them.

**Comment by jplatte** [comment]
Sorry, I meant to say "reference-taking fns". But I checked with my own playground example again and just realized that the reference-taking impl can't possibly apply to `fn(&'static T)`. Sorry for the noise.

**Review by matthewjasper — COMMENTED** [review]
This looks good.
Can you bless the nll compare mode tests if you haven't already?
Presumably this is now a hard error:
```
trait X {}

impl X for fn(fn(&())) {}
impl X for fn(fn(&'static ())) {}
```
Can a test for this be added.

File: src/test/ui/type-alias-impl-trait/issue-57611-trait-alias.stderr [hunk 1]
```diff
@@ -1,23 +1,12 @@
...
 LL |     type Bar = impl Baz<Self, Self>;
-   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected signature of `for<'r> fn(&'r X) -> _`
-...
-LL |         |x| x
-   |         ----- found signature of `fn(_) -> _`
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ one type is more general than the other
    |
-   = note: the return type of a function must have a statically known size
+   = note: expected type `std::ops::FnOnce<(&X,)>`
+              found type `std::ops::FnOnce<(&X,)>`
```
[the remark this answers is not in the document]
Comment (nikomatsakis) [reply src/test/ui/type-alias-impl-trait/issue-57611-trait-alias.stderr:8]: Hmm good point

**Comment by nikomatsakis** [comment]
@matthewjasper 

The test you suggested

```rust
trait X {}
impl X for fn(fn(&())) {}
impl X for fn(fn(&'static ())) {}
```

is still a warning. Let me see if I think this is correct though... to equate those two functions we get

* `fn(&) <: fn(&'static)`
    * `fn(&'1) <: fn(&'static)`
        * `'static: '1` -- OK, `'1` can be inferred to `'empty`
* `fn(&'static) <: fn(&)`
    * `fn(&'static) <: fn(&'!1)`
        * `'1: 'static` -- error from leak-check, because placeholder can reach a U0 region

So, we get an error, but the error is caused by the leak-check, so I think a warning is correct?

Unless I messed something up there... I do remember wanting to create a comprehensive serious of tests around coherence and regions and fns, I can't remember if I did, maybe I'll go back and try to do that. Gotta do get dinner ready now though!

**Comment by matthewjasper** [comment]
I meant

```rust
impl X for for<'a> fn(fn(&'a ())) {}
impl X for fn(fn(&'static ())) {}
```

which for some reason is what I thought what I wrote desugared to.

But that's already a hard error, although this isn't:

```rust
impl X for for<'r> fn(fn(&'r ())) {}
impl<'a> X for fn(fn(&'a ())) {}
```

**Comment by nikomatsakis** [comment]
@matthewjasper ok -- that example does indeed give a hard error with this PR (but not with nightly). It even cites #56105 =) 

```
error[E0119]: conflicting implementations of trait `Trait` for type `for<'r> fn(fn(&'r ()))`:
  --> coherence-fn-covariant-bound-vs-static.rs:16:1
   |
15 | impl Trait for for<'r> fn(fn(&'r ())) {}
   | ------------------------------------- first implementation here
16 | impl<'a> Trait for fn(fn(&'a ())) {}
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ conflicting implementation for `for<'r> fn(fn(&'r ()))`
   |
   = note: this behavior recently changed as a result of a bug fix; see rust-lang/rust#56105 for details
```

**Comment by nikomatsakis** [comment]
Added tests and rebased.

@bors r=matthewjasper

**Comment by nikomatsakis** [comment]
I see that @matthewjasper didn't explicitly say r=me so I'll let them add an explicit comment if desired =)

**Comment by matthewjasper** [comment]
Were there no changes to the NLL compare mode tests?
r=me with all of the tests blessed

**Comment by nikomatsakis** [comment]
I thought I ran with compare-mode=nll but I guess not. Running bless revealed that NLL accepts the old-lub-glb-hr.rs test instead of giving an error, as I expected. I'm not entirely sure why this is the case, have to investigate a bit further I suppose.

File: src/librustc_infer/infer/higher_ranked/mod.rs [hunk 2]
```diff
@@ -81,7 +79,12 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
     where
```
Comment (matthewjasper) [inline src/librustc_infer/infer/higher_ranked/mod.rs:79]: @lcnr had some questions about this function. The doc comment appears to be outdated.

**Comment by matthewjasper** [comment]
where does this leave #60471 #48049 ?

**Comment by nikomatsakis** [comment]
A few notes:

I investigated what is going on with the lub-glb-hr test that is passing on NLL. I still need to dig a bit more but what I think is happening is this:

In the regular type check, the coercion triggers the LUB code, which enforces equality

https://github.com/rust-lang/rust/blob/73558160933b2764ed9a84b1b2b647e128eac3f8/src/librustc_infer/infer/lub.rs#L96-L99

This winds up selecting the `a` type, somewhat arbitrarily, as the "LUB" of the two types (after enforcing that they are equal):

https://github.com/rust-lang/rust/blob/73558160933b2764ed9a84b1b2b647e128eac3f8/src/librustc_infer/infer/lub.rs#L100

This ultimately gives an error from region checking, one which is basically independent from this code. The only way that this PR affects things is that by removing the leak-check from this pathway we do potentially alter the flow of the coercion, in that I think that the `lub` operation used to result in an `Err` eagerly and now it is coming up in the region checking.

I think what happens is that we wind up with the more general type as the "coercion target". We then generate MIR that has a series of `a <- b` assignments where `b` are the specific types and `a` is this coercion target. These path region checking because, indeed, there is a more general type here, we were just having trouble finding it in our older code.

However, this suggests that sometimes we would pick as our coercion target a *less general type*, and I wasn't quite able to make that happen. I thought for example that changing the order of the match arms might do it but it didn't seem to. I'm not sure why not yet, which is why a bit more investigation might be warranted.

**Comment by nikomatsakis** [comment]
Regarding the question that @matthewjasper raised:

> Where does this leave #60471 #48049 ?

#60471 is lazy normalization. I believe that this PR kind of "unblocks" more progress there. We should be able to explore what happens when we stop eagerly normalizing as much and start generating obligations to normalize later. We will hit some cycle issues (at least that's what's happened in the past) but there are some possible solutions. 

#48049 is "chalk-ification". This PR moves us closer to the model that chalk has, in that chalk doesn't have a notion of a leak-check, so it's helpful there. However, chalk still has a more extreme position than rustc does, in that rustc (with this PR) is able to make use of *some amount* of information about higher-ranked things when selecting impls, and thus able to discriminate between `fn(T)` and `fn(&U)`, which chalk couldn't do. Chalk will either have to grow this ability or else we have to remove it from rustc. I consider that "future work", but presently I lean towards both chalk/rustc adopting roughly the model of this PR, but enhanced somewhat to account for implied bounds.

**Comment by nikomatsakis** [comment]
Huh, digging into this a bit more deeply, I see what *seems* to be a bug in the NLL universe solver, such that it accepts a fn that returns one of its two arguments

```
for<'a> fn('a, 'a) -> 'a 
```

as a subtype of a fn that returns its first argument

```
for<'a, 'b> fn('a, 'b) -> 'a
```

This seems clearly false. It doesn't happen, oddly enough, if I run with a test-case that just targets this assignment, but it *does* happen when the MIR is generated from a match. Trying to figure out what's going on.

**Comment by nikomatsakis** [comment]
*Ah*, I think I see the bug, and oof it was a subtle one. There is some code in the NLL type relating logic that says `if a == b { return }` but that is incorrect if either `a` or `b` involves escaping bound variables. This is because of the "lazy substitution" optimization that the NLL type relater uses -- those escaping bound variables may wind up replaced with different things once that is done.

**Comment by nikomatsakis** [comment]
Unfortunately (or fortunately?) this does have the implication that the NLL solver accepts the old-lub-glb-hr.rs test depending on the order of the match arms: basically, does luck of the draw wind up getting the coercion code to pick the more general case?

I don't know that there's much to be done about that just now, and it is not (yet) exposed to users, but we probably want to add it to some tracking issue relating to the overall NLL transition.

**Comment by nikomatsakis** [comment]
@bors rollup=never

I'm disabling rollups not because I know of anything in particular but because this changes a fundamental enough bit that I suspect some complication or other could arise, and bisection would be easier this way.

**Comment by QuineDot** [comment]
The lint should be updated to reflect the current direction of the language.  Namely, [this:](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=51d20b1fd3f047b6faa506859dbc6815)
```rust
trait Trait {}
impl<T> Trait for fn(&T) { }
impl<T> Trait for fn(T) { } 
```
Says
```
  = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
```
But as I understand it, [the goal is to support that indefinitely.](https://github.com/rust-lang/rust/pull/72493)

**Pull request merged**